sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2548,6 +2548,21 @@ class DataFrameSuite extends QueryTest
       assert(df.schema === new StructType().add(StructField("d", DecimalType(38, 0))))
     }
   }
+
+  test("SPARK-31990: preserves the input order of colNames in dropDuplicates") {
+    val df = Seq((1, 2, 3, 4, 5), (1, 2, 3, 4, 5)).toDF("c", "e", "d", "a", "b")
+    val inputColNames = Seq("c", "b", "c", "d", "b", "c", "b")
```

Comment:
@xuanyuanking Do you have a test input seq to reproduce this issue? I checked if this test suite could fail w/o the hot-fix, but it passed...

Reply:
?

Reply:
I just couldn't find a test query to reproduce the issue described in the JIRA. I think @xuanyuanking can provide it because he find this issue...

Reply:
You may want to find the case where `toSet.toSeq` produces different order than `distinct`.

And then you may need to craft a streaming query doing dropDuplicate with having grouping key just you found, and run the query with Spark 2.4.x (where SPARK-31292 is not applied), and store checkpoint, and run the query with Spark 3.x (where SPARK-31292 is applied) with restoring checkpoint.

There're a couple of test suites doing that - one of example is https://github.com/apache/spark/blob/8282bbf12d4e174986a649023ce3984aae7d7755/sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala#L880-L952

Reply:
Technically, as you're implementing the same as `distinct` does, it doesn't fix the issue for restoring from old checkpoint. Both Spark 3.x and your patch will fail. That's what I commented in JIRA - "do the right fix" vs "leave it as it is".

Reply:
BTW, @HeartSaVioR . Is there a test case failure using the same Spark version checkpointing? I'm curious if this only occurs between different Spark versions.

Reply:
Thanks for adding a new UT here. Since this issue was found when investigating the test failure in https://github.com/apache/spark/pull/28707#issuecomment-639861273,
how about reusing the UT `default config of includeHeader doesn't break existing query from Spark 2.4` in `KafkaMicroBatchV2SourceSuite`? I think we don't need to add a new UT for this regression after #28707. That is to say after #28707 is merged, if we don't do the fix, the mentioned UT will fail.